src/tools/gdb/plan_dump.py

```diff
@@ -0,0 +1,590 @@
+
+	def __init__(self, node, rtableMap, typ):
+		self.scan = node.cast(Scan.gdb_type)
+		self.__rtableMap = rtableMap
+		self.__typ = typ
+
+	def print_node(self, indexid):
+		indexInfo = ""
+		if indexid is not None:
+			indexInfo = " (used index, indexoid %s)" % str(indexid)
+		id = int(self.scan["scanrelid"])
+		relanme = self.__rtableMap[id]
+		return "%s on %s%s" % (self.__typ, relanme, indexInfo)
+
+class PlanDumperCmd(gdb.Command):
+	"""Print the plan nodes like pg explain"""
+
+	def __cleanup__(self):
+		self.__result = ""
+		self.__tabCnt = 0
+		self.__curPlanName = None
+		self.__currentSlice = 0
+		self.__pstmt = None
+		self.__state = None
+		self.__rtableMap = {}
+		self.__reloidMap = {}
+
+	def __init__(self):
+		super(PlanDumperCmd, self).__init__("plan_dump_cmd", gdb.COMMAND_USER)
+		self.__cleanup__()
+
+	def walk_initplans(self, plans, sliceTable):
+		"""
+		This function (and walk_subplans) is an analog of ExplainSubplans
+		from explain.c (for init plans), but instead of working with the PlanStates (like
+		original function does) of plan (the EXPLAIN has all states of
+		plan nodes starting from root), this function can't rely on PlanStates,
+		because coredump's from segments may contain only part planstate
+		(queryDesc->planstate) nodes of the original plan
+		(queryDesc->plannedstmt->planTree), so this function works with the
+		Plan nodes, and it's processing differs (in case of subPlans (not init plans
+		the PlanState nodes have the subPlan field, while the Plan nodes does not))
+		"""
+		head = plans["head"]
+		saved_slice = self.__currentSlice
+		while head != 0:
+			sp = head["data"]["ptr_value"].cast(SubPlan.gdb_type)
+			if sliceTable != 0 and sp["qDispSliceId"] > 0:
+				self.__currentSlice = List.list_nth(sliceTable["slices"], sp["qDispSliceId"]).cast(Slice.gdb_type)
+
+			self.__curPlanName = sp["plan_name"]
+			sp_plan = List.list_nth(self.__pstmt["subplans"], int(sp["plan_id"]) - 1).cast(Plan.gdb_type)
+			self.walk_node(sp_plan)
+			head = head["next"]
+		self.__currentSlice = saved_slice
+
+	def walk_subplans(self, plannode, sliceTable):
+		"""
+		Analog of ExplainSubPlans (only for subplans initPlans are handle by
+		function above, unlike original  verison). Works with Plan nodes unlike
+		original version which works with PlanState.
+		"""
+		if plannode["qual"] != 0:
+			self.walker(plannode["qual"], sliceTable)
+		if plannode["targetlist"] != 0:
+			self.walker(plannode["targetlist"], sliceTable)
+
+		# Processing of other node types should be added later.
+		tag = str(plannode["type"])
+		if tag == "T_Result":
+			resNode = plannode.cast(Result.gdb_type)
+			if resNode["resconstantqual"] != 0:
+				self.walker(resNode["resconstantqual"], sliceTable)
+		elif tag in ["T_NestLoop", "T_MergeJoin", "T_HashJoin"]:
+			joinQual = plannode.cast(Join.gdb_type)["joinqual"]
+			if joinQual != 0:
+				self.walker(joinQual.cast(Node.gdb_type), sliceTable)
+			# and node specific quals should be added here
```

Comment:
Maybe it's worth to add processing of subplans, which wasn't processed. I mean that maybe it's worth to maintain `quertyDesc->plannedStmt->subplans` list (which subplans was processed by plan_id on the init_plan phase, or by the quals), and suplans which wasn't processed - print them with the some warning message, like "this case wasn't covered by script"